### DISCUSSION TITLE
Issues with Atomic Guards left behind after documents deleted

### DISCUSSION THREAD

**Discussion by mcrio** [ravendb/ravendb#16460] [Q&A] [answered]
Hi, let's consider:

- Cluster wide transaction
- Atomic guards enabled
- Semantic document id: `foo/bar/baz` produces compare exchange value `rvn-atomic/foo/bar/baz`

When we delete the document the atomic guard will remain and should be deleted automatically by RavenDB some time in future.

The problem is that until that atomic guard is removed we cannot insert a new document with the same id `foo/bar/baz` as it'll not match the existing atomic guard index. On my development instance, 20 minutes after document deletion the atomic guard is still not removed.

Test case: 

```csharp
public class AtomicGuardDocumentRemovalTest : RavenTestDriver
{
    [Fact]
    public void ShouldRemoveExistingDocumentAndAddAnotherUsingTheSameSemanticId()
    {
        var store = GetDocumentStore();
        const string semanticId = "articles/my-article";

        // Insert document
        using (var session = store.OpenSession(new SessionOptions()
               {
                   TransactionMode = TransactionMode.ClusterWide,
                   DisableAtomicDocumentWritesInClusterWideTransaction = false,
               }))
        {
            session.Store(new Article { Id = semanticId, });
            session.SaveChanges();
        }

        WaitForIndexing(store);
        
        // Delete document
        using (var session = store.OpenSession())
        {
            var existing = session.Load<Article>(semanticId);
            Assert.NotNull(existing);
            
            session.Delete(existing);
            session.SaveChanges();
        }
        
        WaitForIndexing(store);
        
        // Make sure document is deleted 
        using (var session = store.OpenSession())
        {
            var deletedDocument = session.Load<Article>(semanticId);
            Assert.Null(deletedDocument);
        }

        // you can uncomment below, it will still fail
        //Thread.Sleep(TimeSpan.FromSeconds(10));
        
        // Insert new document with same id of deleted document
        using (var session = store.OpenSession(new SessionOptions()
               {
                   TransactionMode = TransactionMode.ClusterWide,
               }))
        {
            session.Store(new Article { Id = semanticId, });
            session.SaveChanges();
        }
    }

    public class Article
    {
        public string Id { get; set; }
    }
}

```

**Comment by garayx** [comment, marked as answer]
Hi,

Thanks for the failing test, 

The deletion of document the was created along with atomic guard, should be done is a `ClusterWide` session as well.

I any case there is a open ticket to remove the atomic guard in the regular session as well.

https://issues.hibernatingrhinos.com/issue/RavenDB-19629/Atomic-guard-is-out-of-sync-if-it-was-deleted-in-a-regular-tx

**Reply by mcrio** [reply]
Hi @garayx 

Thanks, I was testing something and removed the documents from the studio.

The feature from the ticket will be a huge helper as it'll prevent scenarios where people will figure out this problem to late. Also the docs seem to be missing the info about a cluster-wide deletion.

~~Anyway, I modified the test case so the deletion is a cluster-wide transaction but the atomic guard still remains, even when adding a 10 seconds delay. Please can you check as well, or did I misunderstand something?~~

**Reply by mcrio** [reply]
NOTE: Please disregard the failing test mentioning. I miscoded the test. It does work when it's a cluster wide transaction.

**Comment by ayende** [comment]
https://github.com/ravendb/ravendb/pull/16909
